watchtower/wtdb/range_index.go

```diff
@@ -0,0 +1,350 @@
+package wtdb
+
+import (
+	"sync"
+)
+
+// RangeIndex can be used go keep track of which numbers have been added to a
```

Comment:
I think I'm missing the motivation a bit here, in which case would there be a series of disparate gaps in what we've uploaded to the tower? 

In other words, afaict, even if we're down for some time, we won't have gaps in _prior_ heights uploaded unless the tower actually deleted information. 

In the case where we switching to a tower and want to know the extent that we've re-uploaded our prior states, we'd still only need to track a single height (the final height that we've uploaded).

Reply:
> in which case would there be a series of disparate gaps in what we've uploaded to the tower?

If we have more than one tower, we might switch to a session with a different tower for a bit if the first one goes down. Also on startup we just pick a random session to start sending updates on. In the idea case, yes there would not be gaps but we need to account for there being gaps. In future we may even want a more "Round Robin" type thing were we cycle through sessions more often & dont just use one till exhaustion before moving on to the next one.

> In other words, afaict, even if we're down for some time, we won't have gaps in prior heights uploaded unless the tower actually deleted information.

There is nothing that ensures that we send the commitment update heights in a monotonically increasing way on a single session. On startup we just pick a random non-exhausted session to continue on.

> In the case where we switching to a tower and want to know the extent that we've re-uploaded our prior states, we'd still only need to track a single height (the final height that we've uploaded).

Dont think I agree - if we have some of the commitments for a channel backed up to one tower and some to another & then we want remove the one tower, we just want to know the heights we uploaded to that session so that we know which we need to re-upload to the new tower

Reply:
> I thought the queues involved in sending to the tower were FIFO? 

they are FIFO. But what can happen is that: we send backups for heights 1, 2, 3 on session A. Then we restart but this time choose session B as our active session (cause maybe the tower for session A is temporarily down) so we send heights 4, 5, 6 on session B. Now we restart again and choose session A again etc. 
See [here](https://github.com/lightningnetwork/lnd/blob/master/watchtower/wtclient/client.go#L367) where we load all our candidate sessions (map!) that we will iterate through to find one that works

Reply:
> I couldn't find where this logic is - can you point me to it? Or is this something for the future?

I think my comment above answers this. Let me know if it does not

Reply:
oh, I see what you mean. I was mixing up the definition of "monotonically". 
So: 
- session seq number will be monotonically increasing _without_ gaps.
- commitment height of a channel within a session is monotonically increasing and _could_ possible have gaps.